tests/test_handler_rocauc_dist.py

```diff
@@ -31,12 +31,12 @@ def test_compute(self):
             auc_metric.update([y_pred, y])
 
         if dist.get_rank() == 1:
-            y_pred = torch.tensor([[0.2, 0.1], [0.1, 0.5]], device=device)
-            y = torch.tensor([[0], [1]], device=device)
+            y_pred = torch.tensor([[0.2, 0.1], [0.1, 0.5], [0.3, 0.4]], device=device)
+            y = torch.tensor([[0], [1], [1]], device=device)
```

Comment:
if I change line 35 to `y = torch.tensor([[0.], [1.], [1.]], device=device)` I get some error from `all_gather`:
(not 100% sure if it's a bug from our side or from ignite when gathering results with heterogeneous data types)
```
python -m tests.test_handler_rocauc_dist
libc++abi.dylib: libc++abi.dylib: terminating with uncaught exception of type gloo::EnforceNotMet: [enforce fail at ../third_party/gloo/gloo/transport/uv/pair.cc:248] op.nread == op.preamble.nbytes. terminating with uncaught exception of type gloo::EnforceNotMet: [enforce fail at ../third_party/gloo/gloo/transport/uv/pair.cc:248] op.nread == op.preamble.nbytes. 

Process Process-1:
Traceback (most recent call last):
  File "/Users/wenqili/Documents/MONAI/tests/utils.py", line 264, in run_process
    func(*args, **kwargs)
  File "/Users/wenqili/Documents/MONAI/tests/utils.py", line 434, in _call_original_func
    return f(*args, **kwargs)
  File "/Users/wenqili/Documents/MONAI/tests/test_handler_rocauc_dist.py", line 39, in test_compute
    np.testing.assert_allclose(0.66667, result, rtol=1e-4)
  File "/Users/wenqili/Documents/MONAI/monai/handlers/roc_auc.py", line 92, in compute
    _target_tensor = evenly_divisible_all_gather(_target_tensor)
  File "/Users/wenqili/Documents/MONAI/monai/handlers/utils.py", line 79, in evenly_divisible_all_gather
    data = idist.all_gather(data)
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/site-packages/ignite/distributed/utils.py", line 347, in all_gather
    return _model.all_gather(tensor)
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/site-packages/ignite/distributed/comp_models/base.py", line 165, in all_gather
    return self._collective_op(tensor, self._do_all_gather)
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/site-packages/ignite/distributed/comp_models/base.py", line 147, in _collective_op
    tensor = self._apply_op(tensor, device, fn, *args, **kwargs)
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/site-packages/ignite/distributed/comp_models/base.py", line 125, in _apply_op
    tensor = fn(tensor, *args, **kwargs)
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/site-packages/ignite/distributed/comp_models/native.py", line 323, in _do_all_gather
    dist.all_gather(output, tensor)
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/site-packages/torch/distributed/distributed_c10d.py", line 1439, in all_gather
    work.wait()
RuntimeError: [../third_party/gloo/gloo/transport/uv/unbound_buffer.cc:67] Timed out waiting 60000ms for recv operation to complete

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/wenqili/Documents/MONAI/tests/utils.py", line 264, in run_process
    func(*args, **kwargs)
KeyboardInterrupt

```

Reply:
Hi @wyli ,

Thanks for your test, I tested with your data in NCCL and got same issue, and I also tested with evenly divisible data, got same issue. So I think it's expected behavior of PyTorch `all_gather`, users should not all_gather **data with different type**.

Thanks.

Reply:
we'll run into this case if the user provides labels such as `y = [[0], [1], [0.], [1.]]`, but outside of this distributed context `y = torch.tensor(y)` will automatically make everything floats.
 Because our metrics don't have output datatype guarantee, I think we need something like `data = data.float()`... (but I'm not sure whether this will break any existing feature)

Reply:
Add doc-string in the latest commit to highlight this potential issue.
Thanks.